### PR TITLE
fix(samples/headless-ssr): revisit url management with Search Parameter Manager

### PR DESCRIPTION
**Pull request by y-lakhdar**
https://coveord.atlassian.net/browse/KIT-2945

### TL;DR:
This Pull Request addresses issues with URL encoding in the Search Parameter manager. The UrlManager is no longer suitable due to the need for advanced serialization of search parameters and unnecessary serialization operations. The solution introduces the `CoveoNextJsSearchParameterSerializer` class, which handles search parameters with arrays more effectively. This class generates URLs with a structure similar to Atomic, enhancing the serialization and deserialization of Coveo search parameters.

### Details:
The previous solution did not handle search parameters with arrays properly, especially when dealing with facet values. To address this, the PR reintroduces the `CoveoNextJsSearchParameterSerializer` class, which improves the serialization and deserialization of Coveo search parameters. Unlike the previous approach, this class generates URLs with a structure similar to Atomic, with slight modifications for facet value search parameters.

Example:
Initial implementation URL:
`http://localhost:3001/react?search-state=%7B%22q%22%3A%22test+%22%2C%22f%22%3A%7B%22author-1%22%3A%5B%22BBC+News%22%2C%22Ron+Kettering%22%5D%7D%7D`

New implementation URL:
`http://localhost:3001/react?q=test+&f-author-1=BBC+News&f-author-1=Ron+Kettering`

The Search parameters are now more legible.

In KIT - 2952, the `CoveoNextJsSearchParameterSerializer` class and related utilities will be bundled into the SSR package, simplifying user interactions with the two public methods (`fromUrlSearchParameters` and `fromCoveoSearchParameters`).

### Other changes
- Support for non-Coveo search parameters (e.g. `?q=test&foo=bar`).
- Prevention of redundant history entries for identical facet values.
- Strong coupling between this class and Search Parameters in Headless, to prevent overlooking new search parameters added to Headless.

### CODE REVIEW

File: packages/samples/headless-ssr/common/components/common/search-parameter-utils.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+      return arr1[0] === arr2[0];
+    }
+    if (arr1.every((value) => arr2.includes(value))) {
+      return !arr1.every((value, idx) => arr2.indexOf(value) === idx);
+    }
+  }
+  return false;
+}
+
+function isObject(obj: unknown): obj is object {
```
Comment (olamothe) [inline packages/samples/headless-ssr/common/components/common/search-parameter-utils.ts:163]: There's a `isRecord` function available in `bueno`

**Review by olamothe — APPROVED** [review]
Some comments, but great job in general !

Was not an easy one
